crates/adapters/src/transport/durable/kafka/input.rs

```diff
@@ -0,0 +1,771 @@
+            .chain(type_specific_options.iter())
+            .map(|(o, v)| (o.as_str(), v.as_str()))
+            .collect();
+        for (option, val) in overrides {
+            Self::enforce_option(&mut settings, option, val)
+                .with_context(|| format!("Failed to validate Kafka options for {config_name}"))?;
+        }
+        let group_id = &Uuid::new_v4().to_string();
+        Self::enforce_option(&mut settings, "group.id", group_id)
+            .with_context(|| format!("Failed to validate Kafka options for {config_name}"))?;
+        let default_redpanda_server = default_redpanda_server();
+        Self::set_option_if_missing(&mut settings, "bootstrap.servers", &default_redpanda_server);
+
+        let mut config: ClientConfig = settings
+            .iter()
+            .map(|(&o, &v)| (String::from(o), String::from(v)))
+            .collect();
+        if let Some(log_level) = source.log_level {
+            config.set_log_level(RDKafkaLogLevel::from(log_level));
+        }
+
+        Ok(config)
+    }
+}
+
+impl TryFrom<KafkaDurableInputConfig> for Inner {
+    type Error = AnyError;
+
+    fn try_from(source: KafkaDurableInputConfig) -> Result<Self, Self::Error> {
+        const CONSUMER_SETTINGS: &[(&str, &str)] = &[
+            ("enable.auto.commit", "false"),
+            ("enable.auto.offset.store", "false"),
+            ("auto.offset.reset", "earliest"),
+            ("isolation.level", "read_committed"),
+            ("fetch.wait.max.ms", "0"),
+            ("fetch.min.bytes", "1"),
+            ("enable.partition.eof", "false"),
+        ];
+        let index_consumer_config = Self::kafka_config(
+            &source,
+            &source.consumer_options,
+            CONSUMER_SETTINGS,
+            "consumer",
+        )?;
+
+        let mut reader_consumer_config = index_consumer_config.clone();
+        reader_consumer_config.set("fetch.wait.max.ms", "1000");
+
+        const PRODUCER_SETTINGS: &[(&str, &str)] = &[
+            ("acks", "all"),
+            ("enable.idempotence", "true"),
+            ("batch.size", "1"),
+            ("batch.num.messages", "1"),
+            ("retries", "5"),
+            ("socket.nagle.disable", "true"),
+            ("linger.ms", "0"),
+        ];
+        let mut producer_config = Self::kafka_config(
+            &source,
+            &source.producer_options,
+            PRODUCER_SETTINGS,
+            "producer",
+        )?;
+        producer_config.remove("group.id");
+
+        let index_topic = match source.index_topic {
+            Some(ref s) => s.clone(),
+            None => format!("{}_input-index", source.data_topic),
+        };
+
+        Ok(Inner {
+            index_consumer_config,
+            reader_consumer_config,
+            producer_config,
+            data_topic: source.data_topic.clone(),
+            index_topic,
+            partition: source.partition,
+        })
+    }
+}
+
+/// Durable Kafka input endpoint.
+struct Endpoint(Inner);
```

Comment:
Oops, that's detritus from a series of refactorings. I'll eliminate `Inner`.

Reply:
In fixing it, I realized why it was there. It's because `Endpoint` and `Reader` both want a copy of it. I changed it to `Arc<Inner>`. I'm also going to rename `Inner` to something better.